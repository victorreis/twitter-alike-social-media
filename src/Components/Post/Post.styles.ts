import styled, { CSSObject, DefaultTheme } from 'styled-components';

import { Typography } from '../Typography';

const getPostBorders = ({
  theme,
  compact,
}: {
  theme: DefaultTheme;
  compact: boolean;
}): CSSObject => {
  if (compact) {
    return {
      borderWidth: '1px',
      borderStyle: 'solid',
      borderColor: theme.colors.background.default.light,
      borderRadius: 16,
    };
  }

  return {
    borderBottomWidth: '1px',
    borderBottomStyle: 'solid',
    borderBottomColor: theme.colors.background.default.normal,
  };
};

const getPostContentButtonsContainerBorder = ({
  theme,
  hasChildren,
}: {
  theme: DefaultTheme;
  hasChildren: boolean;
}): CSSObject => {
  return {
    ...(!hasChildren && {
      borderTopWidth: '1px',
      borderTopStyle: 'solid',
      borderTopColor: theme.colors.background.default.normal,
    }),
  };
};

export const PostContainer = styled.div`
  display: flex;
  padding: 16px;
  column-gap: 1rem;

  ${getPostBorders}
`;

export const PostContent = styled.div`
  display: flex;
  flex-direction: column;
  column-gap: 1rem;
  flex: 1;
`;

export const PostContentHeader = styled.div`
  display: flex;
  column-gap: 1rem;
  align-items: center;
`;

export const PostContentTitle = styled(Typography)`
  margin: 0;
`;

export const PostText = styled(Typography)`
  flex: 1;
  margin-bottom: 16px;
`;

export const PostContentButtonsContainer = styled.div<{ hasChildren: boolean }>`
  display: flex;
  flex: 1;
  column-gap: 1rem;
  align-items: center;
  align-self: stretch;
  justify-content: flex-end;
  padding-top: 16px;

  ${getPostContentButtonsContainerBorder};
`;
