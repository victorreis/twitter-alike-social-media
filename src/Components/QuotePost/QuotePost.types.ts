import { TestProps } from '../../Config/Tests/Test.types';
import { QuotePostType } from '../../QuotePost.types';

export interface RequiredQuotePostProps {}

export interface DefaultQuotePostProps {}

export interface OptionalQuotePostProps {
  /**
   * Sets the component styles.
   */
  style?: React.CSSProperties;
}

export type QuotePostProps = RequiredQuotePostProps &
  DefaultQuotePostProps &
  OptionalQuotePostProps &
  TestProps &
  QuotePostType &
  Omit<React.HTMLAttributes<HTMLDivElement>, 'children'>;

export type QuotePostStyleProps = Required<DefaultQuotePostProps>;
