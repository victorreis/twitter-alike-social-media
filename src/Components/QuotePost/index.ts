export * from './QuotePost.types';
export * from './QuotePost';
