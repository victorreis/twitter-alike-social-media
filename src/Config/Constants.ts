export const SHORT_DATE_FORMAT = 'MM-DD-YYYY';
export const EXTENDED_DATE_FORMAT = 'MMMM D, YYYY';
export const POST_CREATOR_MAX_LENGTH = 777;

export const LOCAL_STORAGE = {
  USERS: 'users',
  POSTS: 'posts',
  FOLLOWER_FOLLOWED: 'followerFollowed',
};
