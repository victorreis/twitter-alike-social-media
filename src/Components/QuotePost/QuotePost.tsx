import { TestProps } from '../../Config/Tests/Test.types';
import { Post } from '../Post';
import { QuotePostProps, DefaultQuotePostProps } from './QuotePost.types';

export const quotePostDefaults: Required<DefaultQuotePostProps> &
  Required<TestProps> = {
  testID: 'QuotePost',
};

export const QuotePost: React.FC<QuotePostProps> = (props): JSX.Element => {
  const {
    testID = quotePostDefaults.testID,
    id,
    createdBy,
    originalPost,
    isAvatarClickable,
    onClickAvatar,
    onClickRepost,
    onClickQuotePost,
    ...others
  } = props;

  const handleClickRepost = () => {
    onClickRepost(id);
  };

  const handleClickQuotePost = (
    originalPostId: string,
    quotePostText: string
  ) => {
    onClickQuotePost(originalPostId, quotePostText);
  };

  const handleClickRepostOriginalPost = () => {
    onClickRepost(originalPost.id);
  };

  const handleClickQuotePostOriginalPost = (
    originalPostId: string,
    quotePostText: string
  ) => {
    onClickQuotePost(originalPostId, quotePostText);
  };

  return (
    <Post
      createdBy={createdBy}
      data-testid={testID}
      id={id}
      onClickQuotePost={handleClickQuotePost}
      onClickRepost={handleClickRepost}
      {...others}
      isAvatarClickable={isAvatarClickable}
      onClickAvatar={onClickAvatar}
      type="Post"
    >
      <Post
        compact
        {...originalPost}
        isAvatarClickable={isAvatarClickable}
        onClickAvatar={onClickAvatar}
        onClickQuotePost={handleClickQuotePostOriginalPost}
        onClickRepost={handleClickRepostOriginalPost}
      />
    </Post>
  );
};
