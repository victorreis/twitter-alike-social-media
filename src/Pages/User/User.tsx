import { useState, useEffect } from 'react';

import { PageContainer } from '../../App.styles';
import { PostCreator } from '../../Components/PostCreator';
import { Tab } from '../../Components/Tab';
import { UserDetails } from '../../Components/UserDetails';
import { TestProps } from '../../Config/Tests/Test.types';
import { useRenderPosts } from '../../Hooks/RenderPosts';
import { useShowUserPage } from '../../Hooks/ShowUserPage';
import { UserType } from '../../Models/User.types';
import { PostTypes } from '../../Services/LocalStorageInitializer';
import { postRetrieverService } from '../../Services/PostRetriever';
import { userRetrieverService } from '../../Services/UserRetriever';
import { UserTabBar } from './User.styles';

export const userDefaults: Required<TestProps> = {
  testID: 'User',
};

export const User: React.FC = (): JSX.Element | null => {
  const [user, setUser] = useState<UserType>();
  const [posts, setPosts] = useState<PostTypes[]>([]);
  const { renderPosts } = useRenderPosts({ posts });

  const { urlNickname, handleCloseUserPage } = useShowUserPage({
    nickname: user?.nickname,
  });

  useEffect(() => {
    const loadedUser = userRetrieverService.getByNickname(String(urlNickname));
    if (!loadedUser) {
      handleCloseUserPage();
      return;
    }
    setUser(() => loadedUser);

    const loadedPosts = postRetrieverService.getAllCreatedByUser(loadedUser.id);
    setPosts(() => loadedPosts);
  }, [handleCloseUserPage, urlNickname]);

  if (!user) {
    return null;
  }

  const { name, thumbnailUrl } = user;

  return (
    <PageContainer data-testid={userDefaults.testID}>
      <UserDetails {...user} />

      <PostCreator
        name={name}
        nickname={String(urlNickname)}
        onChange={() => {}}
        thumbnailUrl={thumbnailUrl}
      />

      <UserTabBar>
        <Tab active text="POSTS" />
      </UserTabBar>

      {renderPosts()}
    </PageContainer>
  );
};
