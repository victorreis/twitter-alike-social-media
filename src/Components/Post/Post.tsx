import { useMemo, useRef, useState } from 'react';

import dayjs from 'dayjs';
import { nanoid } from 'nanoid';

import { EXTENDED_DATE_FORMAT } from '../../Config/Constants.config';
import { TestProps } from '../../Config/Tests/Test.types';
import { TypographyVariant } from '../../Theme/Types/Typographies.types';
import { Avatar } from '../Avatar';
import { Button, ButtonSize } from '../Button';
import { PostCreator } from '../PostCreator';
import { Typography } from '../Typography';
import {
  PostContainer,
  PostContent,
  PostContentButtonsContainer,
  PostContentHeader,
  PostContentTitle,
  PostText,
} from './Post.styles';
import { PostProps, DefaultPostProps } from './Post.types';

export const postDefaults: Required<DefaultPostProps> & Required<TestProps> = {
  testID: 'Post',
  compact: false,
};

export const Post: React.FC<PostProps> = (props): JSX.Element => {
  const {
    testID = postDefaults.testID,
    children,
    id,
    text,
    createdAt,
    createdBy,
    reposts,
    quotPosts,
    compact = postDefaults.compact,
    isAvatarClickable,
    onClickAvatar,
    onClickRepost,
    onClickQuotePost,
    style,
    ...others
  } = props;

  const { name, nickname, thumbnailUrl } = createdBy;
  const [showPostCreator, setShowPostCreator] = useState<boolean>();
  const quotePostCreatorRef = useRef<HTMLTextAreaElement>(null);

  const formattedDate = dayjs(createdAt).format(EXTENDED_DATE_FORMAT);

  const userNameVariant: TypographyVariant = compact ? 'h5' : 'h4';
  const userInfoVariant: TypographyVariant = compact
    ? 'subtitle3'
    : 'subtitle2';
  const textVariant: TypographyVariant = compact ? 'body2' : 'body1';
  const buttonSize: ButtonSize = compact ? 'MD' : 'LG';

  const handleClickRepost = () => {
    onClickRepost(id);
  };

  const handleSubmitQuotePost = (quotePostText: string) => {
    setShowPostCreator(() => false);
    onClickQuotePost(id, quotePostText);
  };

  const renderText = useMemo(
    () => () => {
      return text.split('\n').map((line) => (
        <span key={nanoid()}>
          {line}
          <br />
        </span>
      ));
    },
    [text]
  );

  const renderAvatar = () => {
    return (
      <Avatar
        isAvatarClickable={Boolean(isAvatarClickable)}
        name={name}
        nickname={nickname}
        onClickAvatar={onClickAvatar}
        size={compact ? 'SM' : 'MD'}
        thumbnailUrl={thumbnailUrl}
      />
    );
  };

  const handleShowQuotePostCreator = () => {
    setShowPostCreator(() => true);
    if (quotePostCreatorRef.current) {
      quotePostCreatorRef.current.focus();
    }
  };

  const renderQuotePostCreator = () => {
    return (
      showPostCreator && (
        <PostCreator
          ref={quotePostCreatorRef}
          isAvatarClickable={isAvatarClickable}
          name={name}
          nickname={nickname}
          onClickAvatar={onClickAvatar}
          onSubmit={handleSubmitQuotePost}
          thumbnailUrl={thumbnailUrl}
        />
      )
    );
  };

  return (
    <PostContainer
      key={id}
      compact={compact}
      data-testid={testID}
      style={style}
      {...others}
    >
      {!compact && renderAvatar()}

      <PostContent>
        <PostContentHeader>
          {compact && renderAvatar()}

          <PostContentTitle variant={userNameVariant}>{name}</PostContentTitle>
          <PostContentTitle variant={userInfoVariant}>
            @{nickname}
          </PostContentTitle>
          <Typography variant={userInfoVariant}>- {formattedDate}</Typography>
        </PostContentHeader>

        <PostText data-testid={`${testID}_text`} variant={textVariant}>
          {renderText()}
        </PostText>

        {children}

        <PostContentButtonsContainer hasChildren={Boolean(children)}>
          <Button onClick={handleClickRepost} size={buttonSize}>
            Repost
          </Button>
          <Typography variant={textVariant}>{reposts}</Typography>
          <Button
            disabled={showPostCreator}
            onClick={handleShowQuotePostCreator}
            size={buttonSize}
          >
            Quote post
          </Button>

          <Typography variant={textVariant}>{quotPosts}</Typography>
        </PostContentButtonsContainer>

        {renderQuotePostCreator()}
      </PostContent>
    </PostContainer>
  );
};
