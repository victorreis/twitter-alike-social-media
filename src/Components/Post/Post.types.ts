import { TestProps } from '../../Config/Tests/Test.types';
import { PostType } from '../../Post.types';
import { UserType } from '../../User.types';

export interface RequiredPostProps {
  /**
   * User that created the post.
   */
  createdBy: UserType;
}

export interface DefaultPostProps {
  /**
   * Controls when the Post is compact or not.
   */
  compact?: boolean;
}

export interface OptionalPostProps {
  /**
   * Sets the component styles.
   */
  style?: React.CSSProperties;
}

export type PostProps = RequiredPostProps &
  DefaultPostProps &
  OptionalPostProps &
  TestProps &
  Omit<PostType, 'createdBy'> &
  Omit<React.HTMLAttributes<HTMLSpanElement>, 'children'>;
