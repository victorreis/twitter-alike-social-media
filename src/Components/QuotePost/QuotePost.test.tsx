import dayjs from 'dayjs';

import { SHORT_DATE_FORMAT } from '../../Config/Constants';
import {
  renderJestDomCreator,
  renderRTRCreator,
  screen,
} from '../../Config/Tests/GlobalSetup.config';
import { PostType } from '../../Models/Post.types';
import { QuotePostType } from '../../Models/QuotePost.types';
import { UserType } from '../../Models/User.types';
import { QuotePost, quotePostDefaults } from './QuotePost';
import { QuotePostProps } from './QuotePost.types';

describe('quotePost component tests', () => {
  const originalPostUser: UserType = {
    id: 'user123',
    name: 'Victor Reis',
    nickname: 'victor.reis',
    url: '...',
    thumbnailUrl: 'https://i.pravatar.cc/300',
    createdAt: dayjs('01-01-2011', SHORT_DATE_FORMAT).toDate(),
    following: 333,
    followers: 777,
    numberOfPosts: 999,
  };
  const user: UserType = {
    id: 'user456',
    name: 'Another User',
    nickname: 'another.user',
    url: '...',
    thumbnailUrl: 'https://i.pravatar.cc/300',
    createdAt: dayjs('01-01-2011', SHORT_DATE_FORMAT).toDate(),
    following: 33,
    followers: 77,
    numberOfPosts: 999,
  };
  const originalPost: PostType = {
    type: 'Post',
    id: 'post123',
    text: `first line\nsecond line\nthird line`,
    createdAt: dayjs('01-01-2011', SHORT_DATE_FORMAT).toDate(),
    createdBy: originalPostUser,
    reposts: 9,
    quotPosts: 99,
  };
  const quotePost: QuotePostType = {
    type: 'QuotePost',
    originalPost,
    id: 'post123',
    text: `first line\nsecond line\nthird line`,
    createdAt: dayjs('01-01-2011', SHORT_DATE_FORMAT).toDate(),
    createdBy: user,
    reposts: 9,
    quotPosts: 99,
  };
  const onClickRepost = jest.fn();
  const onClickQuotePost = jest.fn();

  const requiredProps: QuotePostProps = {
    ...quotePost,
    onClickRepost,
    onClickQuotePost,
  };

  const setup = (props?: QuotePostProps) => {
    const renderRTR = () =>
      renderRTRCreator<QuotePostProps>(QuotePost, {
        ...requiredProps,
        ...props,
      });
    const renderJestDom = () =>
      renderJestDomCreator<QuotePostProps>(QuotePost, {
        ...requiredProps,
        ...props,
      });

    return { renderRTR, renderJestDom };
  };

  describe('behavior tests', () => {
    it(`should render the component`, () => {
      expect.assertions(1);
      setup().renderJestDom();
      const testInstance = screen.getByTestId(quotePostDefaults.testID);

      expect(testInstance).toBeTruthy();
    });
  });

  describe('style tests', () => {});

  describe('snapshot tests', () => {
    it(`should render correctly`, () => {
      expect.assertions(1);
      const generatedJson = setup().renderRTR().toJSON();

      expect(generatedJson).toMatchSnapshot();
    });
  });
});
