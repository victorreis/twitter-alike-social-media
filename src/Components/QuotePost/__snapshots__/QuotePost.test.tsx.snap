// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`quotePost component tests snapshot tests should render correctly 1`] = `
<div
  className="PostContainer-twitter-alike-social-media__sc-zy8jam-0 kqSmiT"
  data-testid="QuotePost"
  type="Post"
>
  <div
    className="AvatarContainer-twitter-alike-social-media__sc-ne4tqs-0 fMGQXn"
    data-testid="Avatar"
    size="MD"
  >
    <img
      alt="Another User's avatar photo."
      className="AvatarImage-twitter-alike-social-media__sc-ne4tqs-1 dnbueO"
      data-testid="Avatar_image"
      onClick={[Function]}
      onError={[Function]}
      size="MD"
      src="https://i.pravatar.cc/300"
      title="Another User's avatar photo."
    />
  </div>
  <div
    className="PostContent-twitter-alike-social-media__sc-zy8jam-1 kvzPWr"
  >
    <div
      className="PostContentHeader-twitter-alike-social-media__sc-zy8jam-2 dyHMAc"
    >
      <h4
        className="TypographyContainer-twitter-alike-social-media__sc-1rk4wyq-0 dwIaAW PostContentTitle-twitter-alike-social-media__sc-zy8jam-3 bFlsZm"
        data-testid="Typography"
      >
        Another User
      </h4>
      <h4
        className="TypographyContainer-twitter-alike-social-media__sc-1rk4wyq-0 cCEHbN PostContentTitle-twitter-alike-social-media__sc-zy8jam-3 bFlsZm"
        data-testid="Typography"
      >
        @
        another.user
      </h4>
      <h4
        className="TypographyContainer-twitter-alike-social-media__sc-1rk4wyq-0 cCEHbN"
        data-testid="Typography"
      >
        - 
        January 1, 2011
      </h4>
    </div>
    <span
      className="TypographyContainer-twitter-alike-social-media__sc-1rk4wyq-0 bsZKYT PostText-twitter-alike-social-media__sc-zy8jam-4 kKAkLD"
      data-testid="Post_text"
    >
      <span>
        first line
        <br />
      </span>
      <span>
        second line
        <br />
      </span>
      <span>
        third line
        <br />
      </span>
    </span>
    <div
      className="PostContainer-twitter-alike-social-media__sc-zy8jam-0 fEIXaV"
      data-testid="Post"
      type="Post"
    >
      <div
        className="PostContent-twitter-alike-social-media__sc-zy8jam-1 kvzPWr"
      >
        <div
          className="PostContentHeader-twitter-alike-social-media__sc-zy8jam-2 dyHMAc"
        >
          <div
            className="AvatarContainer-twitter-alike-social-media__sc-ne4tqs-0 kdIcfD"
            data-testid="Avatar"
            size="SM"
          >
            <img
              alt="Victor Reis's avatar photo."
              className="AvatarImage-twitter-alike-social-media__sc-ne4tqs-1 huWvgW"
              data-testid="Avatar_image"
              onClick={[Function]}
              onError={[Function]}
              size="SM"
              src="https://i.pravatar.cc/300"
              title="Victor Reis's avatar photo."
            />
          </div>
          <h5
            className="TypographyContainer-twitter-alike-social-media__sc-1rk4wyq-0 cCEHbN PostContentTitle-twitter-alike-social-media__sc-zy8jam-3 bFlsZm"
            data-testid="Typography"
          >
            Victor Reis
          </h5>
          <h5
            className="TypographyContainer-twitter-alike-social-media__sc-1rk4wyq-0 XqMSr PostContentTitle-twitter-alike-social-media__sc-zy8jam-3 bFlsZm"
            data-testid="Typography"
          >
            @
            victor.reis
          </h5>
          <h5
            className="TypographyContainer-twitter-alike-social-media__sc-1rk4wyq-0 XqMSr"
            data-testid="Typography"
          >
            - 
            January 1, 2011
          </h5>
        </div>
        <span
          className="TypographyContainer-twitter-alike-social-media__sc-1rk4wyq-0 kTmxqV PostText-twitter-alike-social-media__sc-zy8jam-4 kKAkLD"
          data-testid="Post_text"
        >
          <span>
            first line
            <br />
          </span>
          <span>
            second line
            <br />
          </span>
          <span>
            third line
            <br />
          </span>
        </span>
        <div
          className="PostContentButtonsContainer-twitter-alike-social-media__sc-zy8jam-5 DMmfK"
        >
          <button
            className="ButtonContainer-twitter-alike-social-media__sc-1bqjegs-0 hdSPNA"
            data-testid="Button"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            size="MD"
            type="button"
          >
            Repost
          </button>
          <span
            className="TypographyContainer-twitter-alike-social-media__sc-1rk4wyq-0 kTmxqV"
            data-testid="Typography"
          >
            9
          </span>
          <button
            className="ButtonContainer-twitter-alike-social-media__sc-1bqjegs-0 hdSPNA"
            data-testid="Button"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            size="MD"
            type="button"
          >
            Quote post
          </button>
          <span
            className="TypographyContainer-twitter-alike-social-media__sc-1rk4wyq-0 kTmxqV"
            data-testid="Typography"
          >
            99
          </span>
        </div>
      </div>
    </div>
    <div
      className="PostContentButtonsContainer-twitter-alike-social-media__sc-zy8jam-5 gAOtQc"
    >
      <button
        className="ButtonContainer-twitter-alike-social-media__sc-1bqjegs-0 fBnyBV"
        data-testid="Button"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        size="LG"
        type="button"
      >
        Repost
      </button>
      <span
        className="TypographyContainer-twitter-alike-social-media__sc-1rk4wyq-0 bsZKYT"
        data-testid="Typography"
      >
        9
      </span>
      <button
        className="ButtonContainer-twitter-alike-social-media__sc-1bqjegs-0 fBnyBV"
        data-testid="Button"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        size="LG"
        type="button"
      >
        Quote post
      </button>
      <span
        className="TypographyContainer-twitter-alike-social-media__sc-1rk4wyq-0 bsZKYT"
        data-testid="Typography"
      >
        99
      </span>
    </div>
  </div>
</div>
`;
